docs/design/2018-08-29-new-planner.md

```diff
@@ -0,0 +1,411 @@
+  equivalent expressions can be stored in a root Group.
+
+- **Transformation Rule**
+
+  The transform rule is used to transform a logical plan to another equivalent
+  logical plan. It is used to explore all the logically equivalent query plans
+  belonging to an expression group.
+
+- **Implementation Rule**
+
+  The implementation rule is used to implement a logical expression operator to
+  a physical operator. For example, with implementation rules, a logical `Join`
+  operator can be implementated to `HashJoin`/`MergeJoin`/`IndexJoin`, etc.
+
+- **Enforcing Rule**
+
+  Rule to construct an enforcer to satisfy the required physical properties.
+
+- **Pattern**
+
+  Pattern describes a piece of a logical expression. It's a tree-like structure,
+  each node in the tree represents a logical expression operator in the logical
+  expression, the node, or the logical expression operator, is called **Operand**.
+
+- **Operand**
+
+  As disscussed above, the operand represents a logical expression operator. It
+  can be some concrete operator types, for example, `Join`/`Project`/`Filter`.
+
+- **Logical Property**
+
+  Logical properties can be derived from the logical algebra expression, for
+  example: the schema of the expression, the constraints and statistics of the
+  columns in the schema, etc. All the group expressions in the same group shares
+  the same logical property.
+
+- **Physical Property**
+
+  Physical properties depend on the implementation algorithms, or physical
+  operators. Typical physical properties include **order** and **distribution**.
+
+## Proposal
+
+The new planner is composed of 2 phases: exploration and implementation. The
+basic idea comes from the volcano optimizer generator and the optimizations
+mentioned in the cascades project.
+
+### optimization phase: exploration
+
+A typical lifecycle of a SQL Query before this phase is:
+
+```
+SQL Query -> AST -> Logical Plan -> Expression Group
+```
+
+After building the expression group from the input logical plan, the exploration
+optimization phase begins, the target of this phase is to explore all the
+logically equivalent expressions by:
+
+1. Exploring all the equivalent group expressions of each group.
+2. Exploring all the potential groups.
+
+The group and group expression can be defined as:
+
+```go
+type StringSet = map[string]struct{}
+
+type Group struct {
+	equivalents  []*GroupExpr
+	fingerprints StringSet
+	explored     bool
+    ...
+}
+
+type GroupExpr struct {
+	exprNode LogicalPlan
+	children []*Group
+	explored bool
+    ...
+}
+```
+
+The difficult and complex part is how we continually apply some push-down and
+pull-up rules. A simple idea is to traverse the groups and group expressions
+twice: a top-down traverse and a bottom-up traverse. The push-down rules are
+applied during the top-down traverse, the pull-up rules are applied in the
+bottom-up traverse.
+
+But there might be some scenarios that certain push-down rule can also be
```

Comment:
But there might be some scenarios **where a** certain push-down rule can also be
or
But there might be some scenarios **where** certain push-down **rules** can also be